lib/components/app/session-timeout.tsx

```diff
@@ -0,0 +1,118 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+/* eslint-disable react/prop-types */
+import { Button, Modal } from 'react-bootstrap'
+import { connect } from 'react-redux'
+import { FormattedMessage } from 'react-intl'
+import React, { Component } from 'react'
+
+import * as uiActions from '../../actions/ui'
+
+/**
+ * This component makes the current session timeout
+ * by displaying a timeout warning one minute before the timeout,
+ * and by reloading the initial URL if there is no user-initiated
+ * actions within the timeout window.
+ */
+class SessionTimeout extends Component {
+  state = {
+    showTimeoutWarning: false,
+    timeoutObject: null,
+    timeoutStartMillis: 0
+  }
+
+  componentDidMount() {
+    // Wait one second or so after loading before probing for changes
+    // so that initialization actions can complete.
+    setTimeout(this.handleAfterInitialActions, 1500)
+  }
+
+  componentWillUnmount() {
+    // @ts-ignore SessionTimeout is not typed yet
+    clearInterval(this.state.timeoutObject)
+  }
+
+  handleAfterInitialActions = () => {
+    this.setState({
+      timeoutObject: setInterval(this.handleTimeoutWatch, 10000),
+      timeoutStartMillis: new Date().valueOf()
+    })
+  }
+
+  handleTimeoutWatch = () => {
+    // @ts-ignore SessionTimeout is not typed yet
+    const { lastActionMillis, sessionTimeoutSeconds, startOverFromInitialUrl } =
+      this.props
+    if (lastActionMillis > this.state.timeoutStartMillis) {
+      const idleMillis = new Date().valueOf() - lastActionMillis
+      const secondsToTimeout = sessionTimeoutSeconds - idleMillis / 1000
+
+      if (secondsToTimeout < 0) {
+        // TODO: If OTP middleware persistence is enabled,
+        // log out the logged-in user before resetting the page.
+
+        // Reload initial URL (page state is lost after this point)
+        startOverFromInitialUrl()
+      } else {
+        this.setState({
+          // If within a minute of timeout, display dialog, don't otherwise.
+          showTimeoutWarning: secondsToTimeout >= 0 && secondsToTimeout <= 60
```

Comment:
This works as long as sessionTimeoutSeconds is configured to at least a few mins. If it's close to a minute it displays the dialog right away. Maybe dynamically calculate the threshold for displaying the dialog as a fraction of the configured time.

Reply:
I added a shorter warning window for shorter session timeouts (1/3 of the time) in https://github.com/opentripplanner/otp-react-redux/pull/574/commits/090c0aa72dd1b5c23858c77fa101497e47b509c9.